fix(dev): preserve step order during hot reload by reloading all chal…

niyati34 opened the pull request:
…lenges on change/add\n\n- Trigger a full challenges reload when a markdown file changes or is added\n- Use curriculumPath, not a hardcoded english path\n- Mark replacements as reloads to bypass duplicate-id checks in dev\n\nAddresses regression where edited steps moved to the end in maps and nav

Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or Gitpod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #61979

<!-- Feel free to add any additional description of changes below this line -->

Summary

Fixes a regression where editing a single workshop step in dev moved it to the end of the grid and broke next/prev navigation.
Ensures ordering metadata stays consistent during Gatsby hot reload.
Background

After the top‑down curriculum refactor, the source plugin replaced only the edited node on file change, leaving sorting inconsistent in dev.

Changes


- On markdown file change or add, trigger a full challenges reload via the plugin’s source function.
- Use the provided curriculumPath (no hardcoded english path).
- Mark replacements as isReloading to bypass duplicate-id checks during dev.
- File: [gatsby-node.js]


How to test


- Start the client dev server.
- Open a superblock with a step grid.
- Edit any challenge .md in that block.

Verify:

- The step remains in place; grid order is correct.
- Next/prev navigation is unchanged and correct.

Risk/Performance


- Slightly heavier dev reload (full challenges reload) in exchange for correctness. Acceptable for dev; can be optimized later if needed.

Notes
i18n-safe by honoring curriculumPath.

niyati34 committed:
fix(dev): preserve step order during hot reload by reloading all challenges on change/add\n\n- Trigger a full challenges reload when a markdown file changes or is added\n- Use curriculumPath, not a hardcoded english path\n- Mark replacements as reloads to bypass duplicate-id checks in dev\n\nAddresses regression where edited steps moved to the end in maps and nav

ilenia-magoni commented:
> Risk/Performance
> 
> * Slightly heavier dev reload (full challenges reload) in exchange for correctness. Acceptable for dev; can be optimized later if needed.

the previous version, before https://github.com/freeCodeCamp/freeCodeCamp/pull/61459, managed to only recreate the single challenge (or that was the log shown in the terminal), is that not possible?

and if that is not possible, is it necessary to recreate everything? maybe a smaller batch? like just that block?

niyati34 committed:
perf(dev): reload only the affected block on change/add to preserve order with less work\n\n- Recreate all challenges in the edited block instead of full reload\n- Keeps challengeOrder consistent while reducing dev reload cost\n- Uses curriculumPath; marks replacements as reloads to avoid duplicate-id errors

niyati34 commented:
> > Risk/Performance
> > 
> > * Slightly heavier dev reload (full challenges reload) in exchange for correctness. Acceptable for dev; can be optimized later if needed.
> 
> the previous version, before #61459, managed to only recreate the single challenge (or that was the log shown in the terminal), is that not possible?
> 
> and if that is not possible, is it necessary to recreate everything? maybe a smaller batch? like just that block?

- Great point — a full reload was heavier than needed. I’ve updated the plugin to reload only the affected block on file change/add, which keeps challengeOrder consistent without rebuilding everything.
- Recreating just the edited node turned out brittle with Gatsby’s dev data layer (ordering occasionally drifted). Block-scoped reload is a safe middle ground and mirrors where ordering is defined (meta.challengeOrder).
- If we still want to try single-node reload later, I can prototype it behind a flag as a follow‑up.

huyenltnguyen reviewed (commented):
Please help address the lint issue.

ojeytonwilliams commented on tools/client-plugins/gatsby-source-challenges/gatsby-node.js:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable import/no-unresolved */
```
```suggestion
```
We need this one.

ojeytonwilliams reviewed (changes requested):
Hi @niyati34, thanks for taking this on.

It shouldn't be necessary to recreate all the files.  I've not had chance to confirm, but it's likely just that this relied on properties that are no longer returned by `replaceChallenge`.

If you take a look at https://github.com/freeCodeCamp/freeCodeCamp/pull/62056 you can see a similar bug with a similar fix. In that case the page was not getting regenerated because `superBlock` was missing from the object returned by `replaceChallenge` and the fix was simply to add it back.

niyati34 committed:
fix(dev): single-node hot reload with restored ordering metadata

- Switch from block reload to single-challenge replacement as suggested in review
- Add missing superBlock, order, superOrder metadata to prevent challenges jumping to end
- Inspired by fix in #62056 - missing properties cause Gatsby page regeneration issues
- Keep eslint disable for import/no-unresolved as requested
- Remove outdated comment about full-reload paths
- Fix env.json import to use process.env fallback

niyati34 committed:
Merge upstream/main: resolve conflicts and align with array-based API

niyati34 committed:
fix: add missing eslint-disable comment

niyati34 committed:
fix: remove unused eslint-disable comment

- The eslint-disable import/no-unresolved was flagged as unused by linter
- Current imports do not trigger import/no-unresolved rule
- Original file in main branch also doesn't have this comment
- Fixes CI lint failure

ojeytonwilliams reviewed (approved):
> If there’s a specific import you’re seeing that needs it, I can scope-disable just that line.

No, I wasn't thinking straight. You were absolutely right to remove it, thanks!

This looks great, nice work.

The only thing I can see that could be worth doing is to create a helper function and call it on 'add' and on 'change', since the code is identical now. If you want to make that change, please do, but we can merge regardless.

niyati34 committed:
refactor: extract duplicate challenge update logic into helper function

- Create handleChallengeUpdate() helper for both 'add' and 'change' events
- Eliminates code duplication as suggested by @ojeytonwilliams
- Standardizes error handling and logging format

niyati34 commented:
> > If there’s a specific import you’re seeing that needs it, I can scope-disable just that line.
> 
> No, I wasn't thinking straight. You were absolutely right to remove it, thanks!
> 
> This looks great, nice work.
> 
> The only thing I can see that could be worth doing is to create a helper function and call it on 'add' and on 'change', since the code is identical now. If you want to make that change, please do, but we can merge regardless.

Thanks for the feedback! You're absolutely right about the code duplication. 

✅ **Refactored:** Extracted the duplicate logic into a `handleChallengeUpdate()` helper function that both 'add' and 'change' events now use.

**Benefits:**
- Eliminates ~27 lines of duplicate code
- Standardizes error handling and logging format  
- Makes the code more maintainable

The PR should be ready to merge now! 🎉

ojeytonwilliams reviewed (approved):
LGTM :+1: 

Thanks again.

niyati34 commented:
@huyenltnguyen 
 Hi, I’ve resolved the issue and it’s been pending review for a few days now. Could you please prioritize reviewing it at the earliest?

moT01 reviewed (approved):
LGTM @niyati34 🎉 Thanks for contributing to freeCodeCamp 🎉
